internal/storage/v2/memory/tenant.go

```diff
@@ -55,23 +60,121 @@ func (t *Tenant) storeOperation(serviceName string, operation tracestore.Operati
 func (t *Tenant) storeTraces(traceId pcommon.TraceID, resourceSpanSlice ptrace.ResourceSpansSlice) {
 	t.Lock()
 	defer t.Unlock()
-	if foundTraces, ok := t.traces[traceId]; ok {
-		resourceSpanSlice.MoveAndAppendTo(foundTraces.ResourceSpans())
+	if tracesIndex, ok := t.ids[traceId]; ok {
+		resourceSpanSlice.MoveAndAppendTo(t.traces[tracesIndex].trace.ResourceSpans())
 		return
 	}
 	traces := ptrace.NewTraces()
 	resourceSpanSlice.MoveAndAppendTo(traces.ResourceSpans())
-	t.traces[traceId] = traces
 	// if we have a limit, let's cleanup the oldest traces
 	if t.config.MaxTraces > 0 {
 		// we only have to deal with this slice if we have a limit
 		t.evict = (t.evict + 1) % t.config.MaxTraces
 		// do we have an item already on this position? if so, we are overriding it,
 		// and we need to remove from the map
-		if !t.ids[t.evict].IsEmpty() {
-			delete(t.traces, t.ids[t.evict])
+		if t.traces[t.evict] != nil {
+			delete(t.ids, t.traces[t.evict].id)
 		}
 		// update the ring with the trace id
-		t.ids[t.evict] = traceId
+		t.ids[traceId] = t.evict
+		t.traces[t.evict] = &tracesById{
+			id:    traceId,
+			trace: traces,
+		}
+	}
+}
+
+func (t *Tenant) findTraces(query tracestore.TraceQueryParams) []ptrace.Traces {
+	t.RLock()
+	defer t.RUnlock()
+	traces := make([]ptrace.Traces, 0, query.SearchDepth)
+	for traceId, index := range t.ids {
```

Comment:
you need to iterate in reverse order in order to search latest traces first, starting from something like lastEvicted-1 (test it). And you need to be able to wrap around in the ring buffer.

Reply:
do you have a unit test that validates that search is done latest-first?